dbms/src/Functions/LeastGreatest.h

```diff
@@ -120,5 +121,363 @@ class FunctionVectorizedLeastGreatest : public IFunction
+
+    DataTypePtr getReturnTypeImpl(const DataTypes & arguments) const override
+    {
+        if (arguments.size() < 2)
+            throw Exception(
+                fmt::format("Number of arguments for function {} doesn't match: passed {}, should be at least 2.", getName(), arguments.size()),
+                ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH);
+
+        for (const auto & argument : arguments)
+        {
+            if (!argument->isString())
+            {
+                throw Exception(
+                    fmt::format("argument type not string"),
+                    ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT);
+            }
+        }
+        return std::make_shared<DataTypeString>();
+    }
+
+    void executeImpl(Block & block, const ColumnNumbers & arguments, size_t result) const override
+    {
+        size_t num_arguments = arguments.size();
+        if (num_arguments < 2)
+        {
+            throw Exception(
+                fmt::format("Number of arguments for function {} doesn't match: passed {}, should be at least 2.", getName(), arguments.size()),
+                ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH);
+        }
+        if (collator)
+            executeInternal<true>(block, arguments, result);
+        else
+            executeInternal<false>(block, arguments, result);
+    }
+    template <bool use_collator>
+    void executeInternal(Block & block, const ColumnNumbers & arguments, size_t result) const
+    {
+        size_t num_arguments = arguments.size();
+
+        using impl = StringOperationImpl<least, use_collator>;
+
+        std::vector<const ColumnConst *> const_columns;
+        std::vector<const ColumnString *> string_columns;
+        for (size_t i = 0; i < num_arguments; ++i)
+        {
+            const auto * c = block.getByPosition(arguments[i]).column.get();
+            const auto * c_string = checkAndGetColumn<ColumnString>(c);
+            const ColumnConst * c_const = checkAndGetColumnConstStringOrFixedString(c);
+            if (c_const)
+                const_columns.emplace_back(c_const);
+            if (c_string)
+                string_columns.emplace_back(c_string);
+        }
+
+        String const_res;
+        // for const columns
+        if (!const_columns.empty())
+        {
+            if (const_columns.size() == 1)
+            {
+                const_res = const_columns[0]->getValue<String>();
+            }
+            else
+            {
+                auto res = const_columns[0]->getValue<String>();
+                for (size_t i = 1; i < const_columns.size(); ++i)
+                    impl::constantConstant(const_columns[i]->getValue<String>(), res, collator, res);
+                if (string_columns.empty()) // fill result column
+                {
+                    block.getByPosition(result).column = block.getByPosition(result).type->createColumnConst(const_columns[0]->size(), Field(res));
+                    return;
+                }
```

Comment:
This fast code path can be moved outside this branch for the case `const_columns.size() == 1`.